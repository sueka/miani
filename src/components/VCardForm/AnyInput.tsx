import { Checkbox, Group, Input, TextInput } from '@mantine/core'
import { useMemo } from 'react'
import { useRecoilState } from 'recoil'
import { v4 } from 'uuid'
import anyState from '../../recoil/atoms/any/anyState'
import sharedState from '../../recoil/atoms/sharedState'

interface Props {
  type: string
}

const AnyInput: React.FC<Props> = ({ type }) => {
  const [any, setAny] = useRecoilState(anyState(type))
  const [shared, setShared] = useRecoilState(sharedState(anyState(type).key))
  const inputId = useMemo(v4, [])

  return (
    <Input.Wrapper label={type} labelProps={{ htmlFor: inputId }}>
      <Group gap="xs">
        <Checkbox
          checked={shared}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setShared(event.currentTarget.checked)
          }}
        />
        <TextInput
          id={inputId}
          flex={1}
          value={any}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setAny(event.currentTarget.value)
          }}
        />
      </Group>
    </Input.Wrapper>
  )
}

export default AnyInput
