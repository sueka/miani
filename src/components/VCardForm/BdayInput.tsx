import { Checkbox, Group, Input } from '@mantine/core'
import { DatePicker } from '@mantine/dates'
import React, { useState } from 'react'
import { useRecoilState } from 'recoil'

import toLegacyDate from '../../lib/Temporal/toLegacyDate'
import toPlainDate from '../../lib/Temporal/toPlainDate'
import noYearState from '../../recoil/atoms/bday/noYearState'
import bdayState from '../../recoil/atoms/bdayState'
import sharedState from '../../recoil/atoms/sharedState'

const BdayInput: React.FC = () => {
  const [bday, setBday] = useRecoilState(bdayState)
  const [date, setDate] = useState(bday !== null ? toLegacyDate(bday) : null)
  const [shared, setShared] = useRecoilState(sharedState(bdayState.key))
  const [noYear, setMonthDay] = useRecoilState(noYearState)

  return (
    <Input.Wrapper label="Birth date">
      <Group gap="sm">
        <Checkbox
          checked={shared}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setShared(event.currentTarget.checked)
          }}
        />
        <DatePicker
          allowDeselect
          date={date ?? undefined}
          onDateChange={(date: Date | undefined) => {
            setDate(date ?? null)
          }}
          value={bday !== null ? toLegacyDate(bday) : null}
          onChange={(newBday: Date | null) => {
            setBday(newBday !== null ? toPlainDate(newBday) : null)
          }}
        />
      </Group>
      <Checkbox
        checked={noYear}
        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
          setMonthDay(event.currentTarget.checked)
        }}
        label="Omits the year of birth of the object the vCard represents."
      />
    </Input.Wrapper>
  )
}

export default BdayInput
