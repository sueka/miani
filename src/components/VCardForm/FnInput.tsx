import { TextInput } from '@mantine/core'
import { useValidatedState } from '@mantine/hooks'
import React, { useLayoutEffect } from 'react'
import { FormattedMessage, useIntl } from 'react-intl'
import { useRecoilState, useRecoilValue } from 'recoil'

import exit from '../../lib/exit'
import r from '../../lib/tags/r'
import { text, textValue } from '../../patterns'
import fnState from '../../recoil/atoms/fnState'
import versionState from '../../recoil/atoms/vCard/versionState'

const FnInput: React.FC = () => {
  const { formatMessage } = useIntl()
  const version = useRecoilValue(versionState)
  const [recoilFn, setRecoilFn] = useRecoilState(fnState)

  const [fn, setFn] = useValidatedState<string>(recoilFn, (value) => {
    switch (version) {
      case '3.0':
        return r`^${textValue}$`.test(value)

      case '4.0':
        return r`^${text}$`.test(value)
    }
  })

  useLayoutEffect(() => {
    setFn(recoilFn)
  }, [setFn, recoilFn])

  return (
    <TextInput
      required
      label={<FormattedMessage defaultMessage="Formatted name" />}
      placeholder="Mr. John Q. Public\, Esq."
      autocomplete="name"
      value={fn.value}
      error={
        !fn.valid && (
          <span
            dangerouslySetInnerHTML={{
              __html:
                version === '3.0'
                  ? formatMessage({
                      defaultMessage:
                        'Should be a <i>text-value</i> on p. 37, RFC 2426.',
                    })
                  : version === '4.0'
                    ? formatMessage({
                        defaultMessage:
                          'Should be a <i>text</i> on p. 10, RFC 6350.',
                      })
                    : exit(),
            }}
          />
        )
      }
      onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
        setFn(event.currentTarget.value)
        setRecoilFn(event.currentTarget.value)
      }}
    />
  )
}

export default FnInput
