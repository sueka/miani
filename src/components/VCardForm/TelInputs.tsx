import { Button, Input, Stack } from '@mantine/core'
import { FormattedMessage } from 'react-intl'
import { useRecoilState } from 'recoil'
import { ulid } from 'ulid'
import telIdsState from '../../recoil/atoms/telIdsState'
import TelInput from './TelInput'

const TelInputs: React.FC = () => {
  const [telIds, setTelIds] = useRecoilState(telIdsState)

  return (
    <Input.Wrapper
      label={<FormattedMessage defaultMessage="Telephone numbers" />}
    >
      <Stack gap="xs">
        {telIds.map((telId) => (
          <TelInput {...{ telId }} />
        ))}
        <Button
          onClick={() => {
            setTelIds((telIds) => [...telIds, ulid()])
          }}
        >
          <FormattedMessage defaultMessage="Add TEL" />
        </Button>
      </Stack>
    </Input.Wrapper>
  )
}

export default TelInputs
