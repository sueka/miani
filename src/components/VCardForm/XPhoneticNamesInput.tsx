import { TextInput } from '@mantine/core'
import { FormattedMessage } from 'react-intl'
import { useRecoilState } from 'recoil'
import xState from '../../recoil/selectors/xState'

const XPhoneticNamesInput: React.FC = () => {
  const [phoneticLastName, setPhoneticLastName] = useRecoilState(
    xState('X-PHONETIC-LAST-NAME'),
  )
  const [phoneticFirstName, setPhoneticFirstName] = useRecoilState(
    xState('X-PHONETIC-FIRST-NAME'),
  )

  return (
    <>
      <TextInput
        label={<FormattedMessage defaultMessage="Phonetic Last Name" />}
        value={phoneticLastName}
        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
          setPhoneticLastName(event.currentTarget.value)
        }}
      />
      <TextInput
        label={<FormattedMessage defaultMessage="Phonetic First Name" />}
        value={phoneticFirstName}
        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
          setPhoneticFirstName(event.currentTarget.value)
        }}
      />
    </>
  )
}

export default XPhoneticNamesInput
