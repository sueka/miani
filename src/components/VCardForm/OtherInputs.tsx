import { Fieldset, Stack, TagsInput } from '@mantine/core'
import { FormattedMessage } from 'react-intl'
import { useRecoilState } from 'recoil'
import anyTypesState from '../../recoil/atoms/anyTypesState'
import AnyInput from './AnyInput'

const OtherInputs: React.FC = () => {
  const [anyTypes, setAnyTypes] = useRecoilState(anyTypesState)

  return (
    <Fieldset
      legend={<FormattedMessage defaultMessage="Others" />}
      px={{ base: 'md' }}
      pb={{ base: 'md' }}
    >
      <Stack gap="xs">
        <TagsInput value={anyTypes} onChange={setAnyTypes} />
        {anyTypes.map((type) => (
          <AnyInput {...{ type }} />
        ))}
      </Stack>
    </Fieldset>
  )
}

export default OtherInputs
