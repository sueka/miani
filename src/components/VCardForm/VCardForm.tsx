import { Stack } from '@mantine/core'
import BdayInput from './BdayInput'
import FnInput from './FnInput'
import NInput from './NInput/NInput'
import OtherInputs from './OtherInputs'
import TelInputs from './TelInputs'

const VCardForm: React.FC = () => (
  <Stack gap="xs">
    <FnInput />
    <NInput />
    <BdayInput />
    <TelInputs />
    <OtherInputs />
  </Stack>
)

export default VCardForm
