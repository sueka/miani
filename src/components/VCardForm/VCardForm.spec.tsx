import { expect, setSystemTime, test } from 'bun:test'
import { MantineProvider } from '@mantine/core'
import { render } from 'react-dom'
import { IntlProvider } from 'react-intl'
import { RecoilRoot } from 'recoil'

import VCardForm from './VCardForm'

// TODO: Setup babel-plugin-formatjs to resume it!
test.skip('<VCardForm>', () => {
  setSystemTime(new Date('2006-01-02T15:04:05Z'))

  render(
    <MantineProvider>
      <RecoilRoot>
        <IntlProvider locale="en">
          <VCardForm />
        </IntlProvider>
      </RecoilRoot>
    </MantineProvider>,
    document.body,
  )

  expect(document.body.innerHTML).toMatchSnapshot()
})
