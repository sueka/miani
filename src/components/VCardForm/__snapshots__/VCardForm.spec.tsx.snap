// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`<VCardForm> 1`] = `"<style data-mantine-styles="classes">@media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}</style><div style="--stack-gap: var(--mantine-spacing-xs); --stack-align: stretch; --stack-justify: flex-start;" class="m-6d731127 mantine-Stack-root"><div class="m-46b77525 mantine-InputWrapper-root mantine-TextInput-root"><label data-required="true" for="mantine-P0-1" id="mantine-P0-1-label" class="m-8fdc1311 mantine-InputWrapper-label mantine-TextInput-label">FN<span aria-hidden="true" class="m-78a94662 mantine-InputWrapper-required mantine-TextInput-required"> *</span></label><p id="mantine-P0-1-description" class="m-fe47ce59 mantine-InputWrapper-description mantine-TextInput-description">Specifies the <mark>formatted</mark> text corresponding to the <mark>name</mark> of the object the vCard represents.</p><div style="--input-margin-top: calc(var(--mantine-spacing-xs) / 2);" data-variant="default" class="m-6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"><input data-variant="default" placeholder="Mr. John Q. Public, Esq." required="" aria-invalid="false" aria-describedby="mantine-P0-1-description" id="mantine-P0-1" class="m-8fb7ebe7 mantine-Input-input mantine-TextInput-input"></div></div><div class="m-46b77525 mantine-InputWrapper-root"><label data-required="true" for="d7060c63-49e8-4d04-abc8-c9c95abfa76a" id="mantine-P0-7-label" class="m-8fdc1311 mantine-InputWrapper-label">N<span aria-hidden="true" class="m-78a94662 mantine-InputWrapper-required"> *</span></label><p id="mantine-P0-7-description" class="m-fe47ce59 mantine-InputWrapper-description">Specifies the components of the <mark>name</mark> of the object the vCard represents.</p><div style="--stack-gap: var(--mantine-spacing-xs); --stack-align: stretch; --stack-justify: flex-start;" class="m-6d731127 mantine-Stack-root"><div style="--sc-shadow: var(--mantine-shadow-xs);" data-initialization="true" data-with-items-borders="true" role="radiogroup" class="m-cf365364 mantine-SegmentedControl-root"><span style="--sc-indicator-width: 0px; --sc-indicator-height: 0px; --sc-indicator-transform: translate(0px, 0px);" class="m-9e182ccd mantine-SegmentedControl-indicator"></span><div data-active="true" class="m-69686b9b mantine-SegmentedControl-control"><input type="radio" name="mantine-P0-12" id="mantine-P0-12-plainText" data-focus-ring="auto" class="m-1714d588 mantine-SegmentedControl-input" value="plainText"><label data-active="true" for="mantine-P0-12-plainText" class="m-1738fcb2 mantine-SegmentedControl-label">Plain text</label></div><div class="m-69686b9b mantine-SegmentedControl-control"><input type="radio" name="mantine-P0-12" id="mantine-P0-12-components" data-focus-ring="auto" class="m-1714d588 mantine-SegmentedControl-input" value="components"><label for="mantine-P0-12-components" class="m-1738fcb2 mantine-SegmentedControl-label">Components</label></div></div><div class="m-46b77525 mantine-InputWrapper-root mantine-TextInput-root"><div data-variant="default" class="m-6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"><input data-variant="default" placeholder="Public;John;Quinlan;Mr.;Esq." required="" aria-invalid="false" id="d7060c63-49e8-4d04-abc8-c9c95abfa76a" class="m-8fb7ebe7 mantine-Input-input mantine-TextInput-input"></div></div></div></div><div class="m-46b77525 mantine-InputWrapper-root"><label for="mantine-P0-23" id="mantine-P0-23-label" class="m-8fdc1311 mantine-InputWrapper-label">BDAY</label><p id="mantine-P0-23-description" class="m-fe47ce59 mantine-InputWrapper-description">Specifies the <mark>birth date</mark> of the object the vCard represents.</p><div style="--group-gap: var(--mantine-spacing-sm); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;" class="m-4081bf90 mantine-Group-root"><div style="--checkbox-color: var(--mantine-color-blue-filled);" data-label-position="right" data-checked="true" class="m-bf2d988c mantine-Checkbox-root m-5f75b09e mantine-Checkbox-root"><div class="m-5f6e695e mantine-Checkbox-body"><div data-label-position="right" class="m-26062bec mantine-Checkbox-inner"><input id="mantine-P0-28" type="checkbox" class="mantine-focus-auto m-26063560 mantine-Checkbox-input"><svg viewBox="0 0 10 7" fill="none" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" class="m-bf295423 mantine-Checkbox-icon"><path d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z" fill="currentColor" fill-rule="evenodd" clip-rule="evenodd"></path></svg></div><div class="m-d3ea56bb mantine-Checkbox-labelWrapper"></div></div></div><div data-calendar="true"><div class="m-30b26e33 mantine-DatePicker-levelsGroup"><div data-month-level="true"><div class="m-730a79ed mantine-DatePicker-calendarHeader"><button type="button" data-direction="previous" tabindex="0" class="mantine-focus-auto m-2351eeb0 mantine-DatePicker-calendarHeaderControl m-87cf2631 mantine-UnstyledButton-root"><svg viewBox="0 0 15 15" fill="none" xmlns="http://www.w3.org/2000/svg" style="width: 45%; height: 45%; display: block;" data-direction="previous" class="m-367dc749 mantine-DatePicker-calendarHeaderControlIcon"><path d="M3.13523 6.15803C3.3241 5.95657 3.64052 5.94637 3.84197 6.13523L7.5 9.56464L11.158 6.13523C11.3595 5.94637 11.6759 5.95657 11.8648 6.15803C12.0536 6.35949 12.0434 6.67591 11.842 6.86477L7.84197 10.6148C7.64964 10.7951 7.35036 10.7951 7.15803 10.6148L3.15803 6.86477C2.95657 6.67591 2.94637 6.35949 3.13523 6.15803Z" fill="currentColor" fill-rule="evenodd" clip-rule="evenodd"></path></svg></button><button type="button" tabindex="0" class="mantine-focus-auto m-f6645d97 mantine-DatePicker-calendarHeaderLevel m-87cf2631 mantine-UnstyledButton-root">January 2006</button><button type="button" data-direction="next" tabindex="0" class="mantine-focus-auto m-2351eeb0 mantine-DatePicker-calendarHeaderControl m-87cf2631 mantine-UnstyledButton-root"><svg viewBox="0 0 15 15" fill="none" xmlns="http://www.w3.org/2000/svg" style="width: 45%; height: 45%; display: block;" data-direction="next" class="m-367dc749 mantine-DatePicker-calendarHeaderControlIcon"><path d="M3.13523 6.15803C3.3241 5.95657 3.64052 5.94637 3.84197 6.13523L7.5 9.56464L11.158 6.13523C11.3595 5.94637 11.6759 5.95657 11.8648 6.15803C12.0536 6.35949 12.0434 6.67591 11.842 6.86477L7.84197 10.6148C7.64964 10.7951 7.35036 10.7951 7.15803 10.6148L3.15803 6.86477C2.95657 6.67591 2.94637 6.35949 3.13523 6.15803Z" fill="currentColor" fill-rule="evenodd" clip-rule="evenodd"></path></svg></button></div><table class="m-cc9820d3 mantine-DatePicker-month"><thead class="mantine-DatePicker-monthThead"><tr class="mantine-DatePicker-weekdaysRow"><th class="m-18a3eca mantine-DatePicker-weekday">Mo</th><th class="m-18a3eca mantine-DatePicker-weekday">Tu</th><th class="m-18a3eca mantine-DatePicker-weekday">We</th><th class="m-18a3eca mantine-DatePicker-weekday">Th</th><th class="m-18a3eca mantine-DatePicker-weekday">Fr</th><th class="m-18a3eca mantine-DatePicker-weekday">Sa</th><th class="m-18a3eca mantine-DatePicker-weekday">Su</th></tr></thead><tbody class="mantine-DatePicker-monthTbody"><tr class="mantine-DatePicker-monthRow"><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="26 December 2005" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">26</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="27 December 2005" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">27</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="28 December 2005" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">28</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="29 December 2005" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">29</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="30 December 2005" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">30</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="31 December 2005" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">31</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="1 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">1</button></td></tr><tr class="mantine-DatePicker-monthRow"><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="2 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">2</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-today="true" aria-label="3 January 2006" tabindex="0" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">3</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="4 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">4</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="5 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">5</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="6 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">6</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="7 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">7</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="8 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">8</button></td></tr><tr class="mantine-DatePicker-monthRow"><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="9 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">9</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="10 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">10</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="11 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">11</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="12 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">12</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="13 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">13</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="14 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">14</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="15 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">15</button></td></tr><tr class="mantine-DatePicker-monthRow"><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="16 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">16</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="17 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">17</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="18 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">18</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="19 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">19</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="20 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">20</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="21 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">21</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="22 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">22</button></td></tr><tr class="mantine-DatePicker-monthRow"><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="23 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">23</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="24 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">24</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="25 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">25</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="26 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">26</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="27 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">27</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="28 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">28</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-weekend="true" aria-label="29 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">29</button></td></tr><tr class="mantine-DatePicker-monthRow"><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="30 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">30</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" aria-label="31 January 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">31</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="1 February 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">1</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="2 February 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">2</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="3 February 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">3</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="4 February 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">4</button></td><td data-with-spacing="true" class="m-8f457cd5 mantine-DatePicker-monthCell"><button type="button" data-outside="true" aria-label="5 February 2006" tabindex="-1" class="mantine-focus-auto m-396ce5cb mantine-DatePicker-day m-87cf2631 mantine-UnstyledButton-root">5</button></td></tr></tbody></table></div></div></div></div><div style="--checkbox-color: var(--mantine-color-blue-filled);" data-label-position="right" data-checked="true" class="m-bf2d988c mantine-Checkbox-root m-5f75b09e mantine-Checkbox-root"><div class="m-5f6e695e mantine-Checkbox-body"><div data-label-position="right" class="m-26062bec mantine-Checkbox-inner"><input id="mantine-P0-83" type="checkbox" class="mantine-focus-auto m-26063560 mantine-Checkbox-input"><svg viewBox="0 0 10 7" fill="none" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" class="m-bf295423 mantine-Checkbox-icon"><path d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z" fill="currentColor" fill-rule="evenodd" clip-rule="evenodd"></path></svg></div><div class="m-d3ea56bb mantine-Checkbox-labelWrapper"><label for="mantine-P0-83" class="m-8ee546b8 mantine-Checkbox-label">Omits the year of birth of the object the vCard represents.</label></div></div></div></div><div class="m-46b77525 mantine-InputWrapper-root"><label for="mantine-P0-87" id="mantine-P0-87-label" class="m-8fdc1311 mantine-InputWrapper-label">TEL</label><p id="mantine-P0-87-description" class="m-fe47ce59 mantine-InputWrapper-description">Specifies the <mark>telephone</mark> number for telephony communication with the object the vCard represents.</p><div style="--stack-gap: var(--mantine-spacing-xs); --stack-align: stretch; --stack-justify: flex-start;" class="m-6d731127 mantine-Stack-root"><button style="--button-color: var(--mantine-color-white);" type="button" class="mantine-focus-auto mantine-active m-77c9d27d mantine-Button-root m-87cf2631 mantine-UnstyledButton-root"><span aria-hidden="true" class="m-a25b86ee mantine-Button-loader"><span style="--loader-size: calc(var(--button-height) / 1.8); --loader-color: var(--button-color);" class="m-b34414df m-5ae2e3c mantine-Loader-root"></span></span><span class="m-80f1301b mantine-Button-inner"><span class="m-811560b9 mantine-Button-label">Add TEL</span></span></button></div></div><fieldset style="padding-right: var(--mantine-spacing-md); padding-left: var(--mantine-spacing-md); padding-bottom: var(--mantine-spacing-md);" data-variant="default" class="m-84c9523a m-e9408a47 mantine-Fieldset-root"><legend class="m-90794832 mantine-Fieldset-legend">Others</legend><div style="--stack-gap: var(--mantine-spacing-xs); --stack-align: stretch; --stack-justify: flex-start;" class="m-6d731127 mantine-Stack-root"><div class="m-46b77525 mantine-InputWrapper-root mantine-TagsInput-root"><div data-variant="default" data-multiline="true" class="m-6c018570 mantine-Input-wrapper mantine-TagsInput-wrapper"><div data-variant="default" class="m-8fb7ebe7 mantine-Input-input mantine-TagsInput-input"><div class="mantine-TagsInput-pillsList m-1dcfd90b mantine-PillGroup-group"><input data-type="visible" aria-haspopup="listbox" autocomplete="off" id="mantine-P0-99" aria-invalid="false" type="text" class="mantine-TagsInput-inputField m-45c4369d mantine-PillsInputField-field"></div></div></div></div><input type="hidden"></div></fieldset></div><div data-portal="true"></div>"`;
